### DISCUSSION TITLE
Reproduce canvas MDN composite "color" mode

### DISCUSSION THREAD

**Discussion by Nboutika** [ImageMagick/ImageMagick#5912] [Help] [answered]
### ImageMagick version

7.1.0

### Operating system, version and so on

Windows 10

### Description

Hello,
I'm doing a project where you have multiple images which make an interface and you can change the color of each component by selecting them, for this I use Konva for the front-end and ImageMagick for the back-end but I use the "color" composite mode and I can't produce a similar result with ImageMagick, I tried multiple things but I think I lack knowledge. In MDN documentation it says "Preserves the luma of the bottom layer, while adopting the hue and chroma of the top layer" so after some researches I tried to change colorspace to HSL in ImageMagick and tried to play with the gamma and the tint but I really couldn't manage to have a similar result do you have any idea on how to have something similar please ? Thank you in advance

I tried: `convert.exe ${inputFile} -set colorspace HSL -fill ${tintColor} -tint 100 -gamma 1.6  ${outputFile}`);` but the results are far from what i'm looking for

**Comment by snibgo** [comment]
As you use IM v7, I suggest using "magick" instead of "convert".

> "Preserves the luma of the bottom layer, while adopting the hue and chroma of the top layer"

This suggests that two input images are required, but your command has only one.

The words "luma", "hue" and "chroma" suggest that HCL may be a suitable colorspace, but you use HSL.

I don't understand what you want to do. Please show sample inputs, with desired outputs. Linking to the "MDN documentation" may also help.

**Comment by Nboutika** [comment]
First thank you very much for your time that's very kind from you, 
 I upgraded my version yesterday, I didn't know that newer versions had "magick" thanks !
Yes sorry I should've be more precise.  So I have an image that has a certain size, then if I want to tint in red for example I'll create an image that has the same size as the one I want to tint and make this new image red and then compose both images. 
I wasn't sure that I could send link here so I didn't : https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/globalCompositeOperation here in the "color" composite 
![image_2022-12-30_163131716](https://user-images.githubusercontent.com/25962594/210086915-a6311883-6d57-4699-ae78-b1196bb7745b.png)

Here we can see the first image is what I get with Konva and the second one is the result when using : `convert input.png -fill #F00C00 -tint 100 output.png` with the same exact input image as we see the result is relatively different, but on ImageMagick blend mode I don't see "color" like there is in MDN so I  tried tint, multiply and other blend modes but I didn't find one that exactly match it
![image](https://user-images.githubusercontent.com/25962594/210087090-087bed77-5615-4c41-b83e-88fbb1de9758.png)

**Comment by snibgo** [comment, marked as answer]
The MDN documentation refers to an operation called "composite" that needs two input images, and a method that describes how the images should be combined.

I think the equivalent to MDN composite with method "color" would be:
```
magick inA.png inB.png -compose Colorize -composite out.png
```

**Reply by Nboutika** [reply]
Thank you for your reply, it is indeed close to the result ! But I have a last problem, the compose is getting applied to transparent area too :
![image](https://user-images.githubusercontent.com/25962594/210227267-f4f2999b-138a-46ae-aa09-f640e708affe.png)

As you can see the green one is the input and the result is good but the transparent area around it is getting the raw color of the second image (inB) I don't know if we can specify a "mode" of composite like dest-in/dest-atop so it only compose where the source image is and not alter transparent areas is it possible ? Thanks again for your help

**Comment by snibgo** [comment]
You don't show your input files or desired output, so advice is difficult.

"`magick -list compose`" will list the compose methods. They are documented at https://imagemagick.org/Usage/compose/ and https://imagemagick.org/script/compose.php .

**Reply by Nboutika** [reply]
I checked the documentation but didnt find what i'm looking for, 
I would like to take this as input:
![in1](https://user-images.githubusercontent.com/25962594/210353800-06255bd9-c3e7-4466-a6b9-12ffce0b3801.png)
and then this one to compose as we said before (to get a result similar to MDN "color" composition)
![in2](https://user-images.githubusercontent.com/25962594/210353883-46373045-35f3-44ef-abd8-5b710ebbc714.png)
And I would like as the output: 
![out](https://user-images.githubusercontent.com/25962594/210353921-194b0604-7653-4a3e-85ef-e651d2e90098.png)
the answer you gave me first gives me the right color, it's really close but the first input has transparent borders so when compositing with the second input it changes to color of the border.
I don't know if we can say how the composite should work, like I want the destination to replace only where the source is and not the transparent area (it's like a dest-in composite I believe, it keeps the transparency) but using Colorize I hope it's clear, you can just compare the output I sent here (which is what I want) and the result of my previous response which has red in the transparent zone
thanks again !

**Reply by fmw42** [reply]
I do not see any transparent borders on the green image

**Comment by fmw42** [comment]
What is your IM version and platform/OS?  What command did you use?  What was your objective?  I do not understand how you went from green to red!

**Reply by Nboutika** [reply]
Hello, thank you for your response, the IM and the platform version is written in the original question: 
"ImageMagick version
7.1.0

Operating system, version and so on
Windows 10"
I think you missed my original question where it says that i'm trying to reproduce the "color" composite mode of MDN (check documentation sent on original question) so it went from green to red due to the composite with the command of the first answer, I used #F00C00 on it to compose so it turned red. 
I described more what i'm looking for in my response to snibgo

**Comment by snibgo** [comment]
>  ... but the first input has transparent borders so when compositing with the second input it changes to color of the border.

As fmw42 says, none of the images you posted has any transparency.

We get some basic information:
```
C:\web\im>%IMG7%magick redRect.png info:
redRect.png PNG 720x190 720x190+0+0 8-bit sRGB 1040B 0.000u 0:00.000

C:\web\im>%IMG7%magick greenRect.png info:
greenRect.png PNG 692x162 692x162+0+0 8-bit sRGB 947B 0.000u 0:00.000

C:\web\im>%IMG7%magick reqOut.png info:
reqOut.png PNG 720x190 720x190+0+0 8-bit sRGB 1032B 0.000u 0:00.000
```
redRect is larger than greenRect, and you want the output to be the same size as redRect.

When compositing two images, the output will be the same size as the first image. So you need a composite with redRect first. `-compose Colorize` will take the luminance from the first image, and chroma and hue from the second. `-compose Luminize` does the opposite.
```
magick redRect.png greenRect.png -compose Luminize -composite out2.png
```
![out2](https://user-images.githubusercontent.com/33812368/210400788-59bd3201-6dc5-4b77-80a2-ce3ff68fc559.png)

**Comment by Nboutika** [comment]
They were the same size but it got altered when I downloaded it from Figma sorry, I found a solution here is the command I use:
`imagemagick2\\magick.exe ${inputFile} ${secondfile} -alpha associate -compose Colorize -composite ${outputFile}` the "smaller" rectangle should be the same size as the second one but it has 16px transparency around it. So with "-alpha associate" I can keep the transparency and it don't get colorized.
Thank you again for your help I will mark your first answer
